# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.
Source -> Compiler -> assembly (all at once) - binary
Source -> Interpreter -> assembly instructions as needed (piece by piece)
python interactive mode like BASH; script mode like a BASH script
 conditional execution -> if condition is true, then do whatever (if, then or if then else)

### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [3]:
help ()



Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> q

You are now leaving help and returning to the Python interpreter.
If you want to ask for help on a particular object directly from the
interpreter, you can type "help(object)".  Executing "help('string')"
has the same effect as typing a particular string at the help> prompt.


### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 
Find out number of miles:

In [6]:
10*1.61

16.1

So 16.1 miles in 43.5 minutes:

In [1]:
43.5/16.1


2.701863354037267

So the average time per mile was 2.70 minutes per mile.
To find MPH

In [2]:
16.1/(43.5/60)

22.206896551724142

So the runner's speed was 22 miles per hour (excessively speedy). 

## Chapter 2
angry debug errors? check that variables are named correctly and not a keyword
can't have spaces in variables, and like in c no capital letters in first word of variable
python math follows PEMDAS (yay!)
division symbol does floor division if it's 1 integer and 1 float -> // solves that or convert all to float ahead of time
concatination works just like matlab does it with matrices, but PYTHON does it with strings; * is repitition of the string
otherwise: really just like matlab or css or c just different syntax so far (because all we are doing is math)

### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1. 9 (integer - b/c floor division)
2. 8.5 (float)
3. 4.0 (float)
4. 11 (integer)
5. ..... (string)

In [9]:
delimiter = '.'
delimiter  * 5

'.....'

### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.


So I really need to make sure I make 2 = 2.0 if I don't want wierd lopity things happening (I need to treat math-y things as floats not integers)

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [19]:
r = 5.0
a = (4.0//3.0)
vol = a*pi*r*r*r
print vol

392.5


This problem (above, the volume one) makes me uneasy. Not sure why. 

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [20]:
(.6*24.95)*60 + 3.0 + 59.0*.75

945.4499999999999

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [24]:
time = 2.0*8.25 + 3.0*(7.2)
after7 = time - 8.0
print after7

30.1


In [46]:
.1*60.0

6.0

You arrive home for breakfast at 7:30:06.  (the second statement evalates seconds)

## Chapter 3
variables are LOCAL
also don't forget dot notation for things INSIDE a module!!!
function input is formally called an argument
and as always, comment your code
if things are syntaxing out: you forgot a colon??
this is pretty much like MATLAB so far but it is more picky about syntax: if colons or indents are left out, CRANKY python :/

### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
                                                                 allen```

In [31]:
def right_justify(s):
    stringlength = len (s)
    addspacelength = 70 - stringlength
    newspacestring = ' '* addspacelength
    newstring = newspacestring + s
    print newstring
right_justify('anna')

    

                                                                  anna


### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [38]:
def printyfunc():
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+'  
printyfunc()

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


(b) Write a function that draws a similar grid with four rows and four columns.

In [44]:

def printyfunc():
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+' 
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+' 
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+' 
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+' 
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|',
    print ' ',
    print ' ',
    print ' ',
    print ' ',
    print '|'
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+',
    print '-',
    print '-', 
    print '-', 
    print '-', 
    print '+', 
    print '-', 
    print '-', 
    print '-', 
    print '-', 
    print '+' 
printyfunc()

+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +


## Chapter 5.1 - 5.7
if and else and elif's all need : to function, otherwise SYNTAX ERRORS
also: if you're doing math mixed with integers and floats, don't want rounding -> use float func to change everything, or if you want it all as integers, use int func
when you get an error: WHERE is it, and WHAT is it? will help you solve (just like CSS or MATLAB)
make sure that recursive functions don't just keep on going forever or else they will hit the recursion limit -> if you're having trouble, draw a stack diagram

### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [53]:
def check_fermat(a,b,c,n):
    import math  
    ctothen =  math.pow(c,n)
    atothen = math.pow(a,n)
    btothen = math.pow(b,n)
    tot = atothen + btothen
    if ctothen == tot:
        print "Holy smokes, Fermat was wrong!"
    else:
        print "No, that doesn't work."
check_fermat(10,12,14,2) # just some random checking numbers

No, that doesn't work.


(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [54]:
def check_fermat2(a,b,c,n):
    import math  
    aint = int(a)
    bint = int(b)
    cint = int(c)
    nint = int(n)
    ctothen =  math.pow(cint,nint)
    atothen = math.pow(aint,nint)
    btothen = math.pow(bint,nint)
    tot = atothen + btothen
    if ctothen == tot:
        print "Holy smokes, Fermat was wrong!"
    else:
        print "No, that doesn't work."
check_fermat2(10,12,14,2) # just some random checking numbers

No, that doesn't work.


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [57]:
def is_triangle(a,b,c):
    if a> b+c:
        print 'No'
    elif b>a+c:
        print 'No'
    elif c>b+a:
        print 'No'
    else:
        print 'Yes'
is_triangle(1,1,2) #more random numbers to check
        
    

Yes


(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [69]:
def is_triangle2():
    a = raw_input('Please input the first stick length')
    b = raw_input('Please input the second stick length')
    c = raw_input('Please input the third stick length')
    afloat = float(a)
    bfloat = float(b)
    cfloat = float(c)
    aint = int(afloat)
    bint = int(bfloat)
    cint = int(cfloat)
    def is_triangle(aint,bint,cint):
        if a> b+c:
            print 'No'
        elif b>a+c:
            print 'No'
        elif c>b+a:
            print 'No'
        else:
            print 'Yes'
    is_triangle(aint,bint,cint)
is_triangle2()

Please input the first stick length12.1
Please input the second stick length1.7
Please input the third stick length2.3
No


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.

Please go over chapter 2, ex 3. I got them numerically right but wasn't sure how 'coded' vs 'calculator style' you wanted so I just did them as I would if I was using MATLAB or wolphram as a tool for solving problems. 
Also you asked for a prompt on chapter 5 ex 4, so I used raw_input but I wasn't sure if you really meant a command line prompt or if you just wanted it to take in the stick lengths as arguements. 